google/cloud/spark_connect/client/proxy.py

```diff
@@ -210,22 +224,33 @@ def _run(self, s):
                 self._port = frontend_socket.getsockname()[1]
             s.release()
             while not self._killed:
-                conn, addr = frontend_socket.accept()
-                # Set a timeout on how long we will allow send/recv calls to block
-                #
-                # The code that reads and writes to this connection will retry
-                # on timeouts, so this is a safe change.
-                #
-                # The chosen timeout is a very short one because it allows us
-                # to more quickly detect when a connection has been closed.
-                conn.settimeout(1)
-                logger.debug(f"Accepted a connection from {addr}...")
-                self._conn_number += 1
-                threading.Thread(
-                    target=forward_connection,
-                    args=[self._conn_number, conn, addr, self._target_host],
-                    daemon=True,
-                ).start()
+                if (
+                    self.callback is not None
+                    and self.callback(self.session_name, self.client_options)
+                    is False
+                ):
+                    self.stop()
+                    logger.error(
+                        f"The serverless session {self.session_name} is not active. Please create a new session"
+                    )
+                    sys.exit(1)
```

Comment:
This would cause the entire Python interpreter to shut down.

That's absolutely not what we want.

I think a better approach would be to add a wrapper method around `self._killed` that also checks this callback if it is defined.

E.G.

```py
    def _is_active(self):
        if self._killed:
            return False
        if self._is_active_callback is not None:
            return self._is_active_callback()
        return True
...
    @property
    def port(self):
        ...

            while self._is_active():
                ...
```